File: sdks/go/cmd/elydora/plugins/claudecode.go
package plugins

import (
	"fmt"
	"path/filepath"
)

// ClaudeCodePlugin manages Claude Code global user hooks.
type ClaudeCodePlugin struct {
	rename renameFunc
}

// ManagesGuardRuntime reports that Claude Code commits its provider guard with
// the audit runtime and user settings document.
func (p *ClaudeCodePlugin) ManagesGuardRuntime() bool {
	return true
}

func requireClaudeHooksEnabled(document *claudeDocument) error {
	if document.hooksDisabled {
		return fmt.Errorf(
			"Claude Code hooks are disabled by disableAllHooks: %s",
			document.filePath,
		)
	}
	return nil
}

// PreflightInstall validates settings and runtime identity before any write.
func (p *ClaudeCodePlugin) PreflightInstall(config InstallConfig) error {
	document, err := readClaudeDocument()
	if err != nil {
		return err
	}
	if err := requireClaudeHooksEnabled(document); err != nil {
		return err
	}
	_, _, err = preflightClaudeInstallation(config, document)
	return err
}

func (p *ClaudeCodePlugin) Install(config InstallConfig) error {
	document, err := readClaudeDocument()
	if err != nil {
		return err
	}
	if err := requireClaudeHooksEnabled(document); err != nil {
		return err
	}
	paths, nodePath, err := preflightClaudeInstallation(config, document)
	if err != nil {
		return err
	}
	hooks, err := removeManagedClaudeHooks(document.hooks, "")
	if err != nil {
		return err
	}
	for _, item := range []struct{ event, script, status string }{
		{"PreToolUse", paths.guardPath, claudeGuardStatusMessage},
		{"PostToolUse", paths.auditPath, claudeAuditStatusMessage},
		{"PostToolUseFailure", paths.auditPath, claudeAuditStatusMessage},
	} {
		hooks[item.event] = append(
			hooks[item.event],
			buildClaudeGroup(nodePath, item.script, item.status),
		)
	}
	rendered, err := renderClaudeDocument(document, hooks)
	if err != nil {
		return fmt.Errorf("render Claude Code user settings: %w", err)
	}
	changes, err := prepareClaudeInstallationChanges(config, paths, rendered)
	if err != nil {
		return err
	}
	if err := writeClaudeChanges(
		changes,
		"Install Claude Code hooks",
		p.rename,
		paths.runtimeRoot,
		paths.agentDirectory,
		filepath.Dir(document.filePath),
	); err != nil {
		return err
	}
	fmt.Printf("Claude Code hooks installed at %s.\n", document.filePath)
	fmt.Println("Claude Code verification: run /hooks and claude doctor.")
	return nil
}

func (p *ClaudeCodePlugin) Uninstall(agentID string) error {
	document, err := readClaudeDocument()
	if err != nil {
		return err
	}
	hooks, err := removeManagedClaudeHooks(document.hooks, agentID)
	if err != nil {
		return err
	}
	rendered, err := renderClaudeDocument(document, hooks)
	if err != nil {
		return fmt.Errorf("render Claude Code user settings: %w", err)
	}
	change, err := prepareRenderedClaudeChange(rendered)
	if err != nil {
		return err
	}
	return writeClaudeChanges(
		[]*fileChange{change},
		"Uninstall Claude Code hooks",
		p.rename,
		"",
		"",
		filepath.Dir(document.filePath),
	)
}

func (p *ClaudeCodePlugin) Status() (PluginStatus, error) {
	configPath, pathErr := claudeSettingsPath()
	entry := SupportedAgents[claudeAgentKey]
	status := PluginStatus{
		AgentName: claudeAgentKey, DisplayName: entry.Name, ConfigPath: configPath,
	}
	if pathErr != nil {
		return status, pathErr
	}
	document, err := readClaudeDocument()
	if err != nil {
		return status, err
	}
	if document.hooksDisabled {
		return status, nil
	}
	contracts, err := claudeRuntimeContracts(document.hooks)
	if err != nil {
		return status, err
	}
	status.HookConfigured = len(contracts) > 0
	if !status.HookConfigured {
		return status, nil
	}
	status.HookScriptExists, err = claudeRuntimeFilesExist(contracts)
	if err != nil {
		return status, err
	}
	status.Installed = status.HookScriptExists
	return status, nil
}

File: sdks/node/src/plugins/claudecode.ts
import type { AgentPlugin, InstallConfig, PluginStatus } from './base.js';
import {
  AGENT_KEY,
  AUDIT_STATUS,
  GUARD_STATUS,
  buildClaudeGroup,
  claudeRuntimeContracts,
  removeManagedClaudeHooks,
  renderClaudeDocument,
} from './claudecode-contract.js';
import {
  commitClaudeInstallation,
  preflightClaudeInstallation,
  prepareClaudeInstallation,
} from './claudecode-installation.js';
import {
  claudeRuntimeFilesExist,
  readClaudeDocument,
  writeClaudeDocument,
} from './claudecode-io.js';
import { SUPPORTED_AGENTS } from './registry.js';

const entry = SUPPORTED_AGENTS.get(AGENT_KEY)!;

function requireEnabledHooks(disabled: boolean, filePath: string): void {
  if (disabled) {
    throw new Error(`Claude Code hooks are disabled by disableAllHooks: ${filePath}`);
  }
}

export const claudecodePlugin: AgentPlugin = {
  managesRuntime: true,

  async preflightInstall(config: InstallConfig): Promise<void> {
    const document = await readClaudeDocument();
    requireEnabledHooks(document.hooksDisabled, document.filePath);
    await preflightClaudeInstallation(config, document.filePath);
  },

  async install(config: InstallConfig): Promise<void> {
    const document = await readClaudeDocument();
    requireEnabledHooks(document.hooksDisabled, document.filePath);
    const paths = await preflightClaudeInstallation(config, document.filePath);
    const hooks = removeManagedClaudeHooks(document.hooks);
    hooks.PreToolUse = [
      ...(hooks.PreToolUse ?? []),
      buildClaudeGroup(paths.guardPath, GUARD_STATUS),
    ];
    hooks.PostToolUse = [
      ...(hooks.PostToolUse ?? []),
      buildClaudeGroup(paths.auditPath, AUDIT_STATUS),
    ];
    hooks.PostToolUseFailure = [
      ...(hooks.PostToolUseFailure ?? []),
      buildClaudeGroup(paths.auditPath, AUDIT_STATUS),
    ];
    const rendered = renderClaudeDocument(document, hooks);
    const prepared = await prepareClaudeInstallation(config, rendered);
    await commitClaudeInstallation(prepared);
    console.log(`  Claude Code hooks: ${document.filePath}`);
    console.log('  Claude Code verification: run /hooks and claude doctor.');
  },

  async uninstall(agentId?: string): Promise<void> {
    const document = await readClaudeDocument();
    const hooks = removeManagedClaudeHooks(document.hooks, agentId);
    await writeClaudeDocument(renderClaudeDocument(document, hooks));
  },

  async status(): Promise<PluginStatus> {
    const document = await readClaudeDocument();
    const contracts = claudeRuntimeContracts(document.hooks);
    const hookConfigured = !document.hooksDisabled && contracts.length > 0;
    const hookScriptExists = hookConfigured
      ? await claudeRuntimeFilesExist(contracts)
      : false;
    return {
      installed: hookConfigured && hookScriptExists,
      agentName: AGENT_KEY,
      displayName: entry.name,
      hookConfigured,
      hookScriptExists,
      configPath: document.filePath,
    };
  },
};
